src/avro/src/decode.rs

```diff
@@ -483,72 +477,81 @@ impl<'a, R: AvroRead> AvroArrayAccess for SimpleArrayAccess<'a, R> {
         let mut des = GeneralDeserializer {
             schema: self.schema,
         };
-        des.deserialize(self.r, d)?;
-        Ok(true)
+        des.deserialize(self.r, d).map(Some)
     }
 }
 
-pub trait AvroDecode {
-    fn record<R: AvroRead, A: AvroRecordAccess<R>>(&mut self, _a: &mut A) -> Result<(), Error> {
+pub trait AvroDecode: Sized {
+    type Out;
+    fn record<R: AvroRead, A: AvroRecordAccess<R>>(self, _a: &mut A) -> Result<Self::Out, Error> {
         bail!("Unexpected record");
     }
 
     fn union_branch<'a, R: AvroRead, D: AvroDeserializer>(
-        &mut self,
+        self,
         _idx: usize,
         _n_variants: usize,
         _null_variant: Option<usize>,
         _deserializer: &'a mut D,
         _reader: &'a mut R,
-    ) -> Result<(), Error> {
+    ) -> Result<Self::Out, Error> {
         bail!("Unexpected union");
```

Comment:
Without digging more I'm not sure when we'd fall back to these default implementations, but will these error messages be enough for a user to understand what went wrong?